9922 Make summaries more accurately account user ops

kian-thompson opened the pull request:
#9922 
[AB#126](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/126)

Currently, we are running heuristic summaries when we observe a certain number of ops come in. The problem is that we consider system and non-system ops to have the same weight, when we really need to be treating non-system ops as more important for creating summaries.
The basic idea is to give non-system ops more weight compared to system ops in terms of determining when to summarize. We don't want to entirely get rid of handling system ops for summaries because this could adversely affect boot times.

kian-thompson committed:
Track count of ops in RunningSummarizer

kian-thompson committed:
Pass op counts to SummarizeHeuristicRunner

kian-thompson committed:
Add logic for weighting different op types

kian-thompson committed:
Reduce default maxOps to 100

kian-thompson committed:
Add comment with concerns

kian-thompson committed:
Pass op counters to heuristicRunner

vladsud commented:
Linking to issue #9922
Hm, it looks more complicated than I anticipated :) Let's chat about it

vladsud commented on packages/runtime/container-runtime/src/containerRuntime.ts:
```diff
@@ -198,8 +198,8 @@ const DefaultSummaryConfiguration: ISummaryConfiguration = {
-    // Snapshot if 1000 ops received since last snapshot.
-    maxOps: 1000,
+    // Snapshot if 100 ops received since last snapshot.
+    maxOps: 100,
```
Let's separate work into multiple smaller PRs. This change will depend on https://github.com/microsoft/FluidFramework/pull/9965 going through first (stop using policy from service), so at the moment this is noop and we should not add it here (for now)

vladsud commented on packages/runtime/container-runtime/src/runningSummarizer.ts:
```diff
@@ -191,10 +193,12 @@ export class RunningSummarizer implements IDisposable {
             case MessageType.ClientLeave:
             case MessageType.ClientJoin:
             case MessageType.Propose: {
+                this.numNonSystemOps--; // Decrement to correct the counters
```
Code before you changes looks wrong - I do not know why it looks only at some system ops.
Please use isSystemMessage() to figure out if message is system or not (it does not have ideal implementation as it uses opt-in vs. ideal out-out, but we will fix it separately)

vladsud commented on packages/runtime/container-runtime/src/runningSummarizer.ts:
```diff
@@ -191,10 +193,12 @@ export class RunningSummarizer implements IDisposable {
             : undefined;
 
     public handleSystemOp(op: ISequencedDocumentMessage) {
+        this.numSystemOps++;
```
This code (separation of handleSystemOp & handleOp looks wrong). Can you please change it to use just one handler? I'd suggest to listen on IDeltaManager's "op" event, but using IDeltaManager.inbound's "op" event is fine too.

vladsud commented on packages/runtime/container-runtime/src/runningSummarizer.ts:
```diff
@@ -206,14 +210,15 @@ export class RunningSummarizer implements IDisposable {
         // Check for enqueued on-demand summaries; Intentionally do nothing otherwise
         if (!this.tryRunEnqueuedSummary()) {
-            this.heuristicRunner?.run();
+            this.heuristicRunner?.run(this.numSystemOps, this.numNonSystemOps);
```
I have a feeling that it's better to move all the logic around tracking ops (including op handler) into this.heuristicRunner.

kian-thompson replied:
I don't know if I'm a huge fan of this idea. I see the heuristic runner as determining, based on the current data, whether a summary should be run or not. In my opinion, it shouldn't also be the one in charge of collecting/manipulating this data.
The only thing it should be "running" is very specific heuristic runners that will self-actuate (ex: idle timer).

vladsud commented on packages/runtime/container-runtime/src/runningSummarizer.ts:
```diff
@@ -333,6 +338,10 @@ export class RunningSummarizer implements IDisposable {
         summarizeReason: SummarizeReason,
         cancellationToken = this.cancellationToken): void
     {
+        // Reset our op counters
```
SummarizeHeuristicData has a bunch of methods to deal with state transitions (recordAttempt, markLastAttemptAsSuccessful). I feel resetting of these counters should be done there, on completion of attempt (assuming we are moving counters and op handler there too).

The only thing that makes me pause is asynchrony. markLastAttemptAsSuccessful() is called long after we unpaused op processing, so that's too late to do so. We can move it to recordAttempt(), but it has two drawbacks - if summary fails, then this reset would be incorrect. Plus looking at where it is called, it's called too late too.

So maybe we should do it here (in a sense of point in time, not necessary place in code), but we should examine if there is any way to "undo" it if summarization failed. I.e. maybe store prior counters and add them (not restore) to current counters when/if summary fails?

BTW, I'd move such logic (or where such logic is initiated) into lockedSummaryAction() itself. As any restore should happen while we still hold a lock.

vladsud commented on packages/runtime/container-runtime/src/summarizerHeuristics.ts:
```diff
@@ -67,20 +67,44 @@ export class SummarizeHeuristicRunner implements ISummarizeHeuristicRunner {
     }
 
-    public run() {
+    public run(numSystemOps?: number, numNonSystemOps?: number) {
```
Optionality of arguments suggests we are not doing right thing for some path (potentially first summary). Can you please take a look?

vladsud commented on packages/runtime/container-runtime/src/summarizerHeuristics.ts:
```diff
@@ -67,20 +67,44 @@ export class SummarizeHeuristicRunner implements ISummarizeHeuristicRunner {
+                // will hold this new "weightedOps" number (100 in this case) and we leave maxOps at 1000.
+                // Another solution could be to instead keep maxOps at 1000 and just divide this number by 10 when
+                // processing the weighted version. But this isn't bullet proof in my opinion.
+                this.idleTimer.clear();
```
Ah, we do cancel timer (I thought we do not when we talked earlier). May I suggest to create a wrapper method to always call this.idleTimer.clear() when we call trySummarize callback? It does happen on 3 paths we have today, but might be easy to miss in the future.

vladsud commented on packages/runtime/container-runtime/src/summarizerHeuristics.ts:
```diff
@@ -67,20 +67,44 @@ export class SummarizeHeuristicRunner implements ISummarizeHeuristicRunner {
     }
 
+    private getWeightedNumOfOps(numSystemOps: number, numNonSystemOps: number): number {
+        return numNonSystemOps + (numSystemOps * 0.1);
```
I'd prefer to move these knobs (1.0 & 0.1) into interface that controls all summary policies. Different workloads would want different policies.

kian-thompson committed:
Switch to old maxOps value

kian-thompson committed:
Refactor a lot of heuristic summarizer code
- Implement new strategy design for more configurability

kian-thompson committed:
Introduce new "start" method
- We shouldn't rely on "run" to start the idle timer. In my opinion, it's better to explicity define a start method instead of expect this side functionality

vladsud commented on packages/runtime/container-runtime/src/summarizer.ts:
```diff
@@ -277,11 +276,8 @@ export class Summarizer extends EventEmitter implements ISummarizer {
-        this.opListener = (error: any, op: ISequencedDocumentMessage) => runningSummarizer.handleOp(error, op);
-        this.runtime.on("batchEnd", this.opListener);
+        this.opListener = (op: ISequencedDocumentMessage) => runningSummarizer.handleOp(op);
+        this.runtime.deltaManager.on("op", this.opListener);
```
Sorry, I gave you somewhat wrong advice.
What you did with ophandler is good for counting ops.

But I missed the fact that we should not run summaries in the middle of a batch. Basically, ops can be grouped into "batches". it's a mechanism to ensure that all related changes are reflected in the model before render happens (or any other async activity). All ops in a batch are always processed in one go, synchronously. 
So earlier code had "batchEnd" handler and summaries were considered only at the end of batch, plus some system ops.
We can summarize after any system op. But, we should not count "summarize" and "summaryAck" ops, as that will result in positive feedback loop / infinitely summarizing summary ops :)

I'd still use more general flow as you have for counting (with exception of excluding explicitly "summarize" and "summaryAck" ops). As for potential trigger to run a summary - I'd think that using micro-task (i.e. Promose.resolve().then) after any non-summary op is a better way to deal with batches. It ensures that we will not break batches (it's a strong guarantee), but it also will find occasionally better point in time to run a summary. I.e.  if we are getting 3 batches each 100 ops long, the best time to summarize after processing all 3 batches, not after just the first one. It's probably noop change today as maxOps = 1000 (so it's very unlikely we would accumulate that many), but we are going to change it to 100, and chances of hitting this condition will increse.

Or we can keep old batchEnd flow + some system ops.

vladsud replied:
It would be nice to take a look if we have a test that validates we are not going into infinite summarization loop due to summary ops

vladsud replied:
One thing to add: It's acceptable not to filter summary ops, but simply disregard 2 system ops in our weighted formula (a though based on suggestion to use OpTracker in other comment).

kian-thompson replied:
> Promose.resolve().then

I'm not familiar with this pattern, but I got the gist of what it's used for reading through some usages. However, I don't understand how using this pattern would help prevent against running the summary multiple times?

vladsud replied:
It will not. It only helps to get on a clean stack (this pattern might be useful to avoid reentrancy as well).
Usually, we add a boolean to track it. See this.pendingReconnect for an example.

vladsud replied:
I'd say without anything else in consideration, running summary on clean stack is better than running it in the middle of op processing pipeline. You never know where it will backfire.

kian-thompson replied:
Not sure if I understand these concepts correctly, but even if we're running on a clean stack wouldn't we potentially be running `runningSummarizer.handleOp` a large number of times that eventually end up doing nothing? I suppose the cost of executing all that code wouldn't be too high, but could potentially be a factor?

kian-thompson replied:
Since it might be difficult to track non-system ops nicely, would it make sense to hook onto the `batchEnd` event again and just track the number of system ops? THen we could do some simple math to determine how many non-system ops there are per summary.

vladsud replied:
Is your concert about perf? There is no way to track just system ops, so if we want to count them, we have to listen to all ops. It not a big perf hit from these handlers, but we should definitely measure. We already were listening for all ops, so you are not adding more cost here.
It feels orthogonal to how / when we decide to run a summary.

vladsud replied:
Is your concert about perf? There is no way to track just system ops, so if we want to count them, we have to listen to all ops. It not a big perf hit from these handlers, but we should definitely measure. We already were listening for all ops, so you are not adding more cost here.
It feels orthogonal to how / when we decide to run a summary.

vladsud commented on packages/runtime/container-runtime/src/summarizerHeuristics.ts:
```diff
@@ -39,46 +45,72 @@ export class SummarizeHeuristicData implements ISummarizeHeuristicData {
+        this.numSystemOps += this.numSystemOpsBefore;
+        this.numNonSystemOps += this.numNonSystemOpsBefore;
+    }
 }
```
I wonder if it's better to do something like that:
```ts
recordAttempt() {
        this.numSystemOpsBefore = this.numSystemOps;
        this.numNonSystemOpsBefore = this.numNonSystemOps;
}

public markLastAttemptAsSuccessful() {
      this.numSystemOps -= this.numSystemOpsBefore;
      this.numSystemOpsBefore = 0;
      this.numNonSystemOps -= this.numNonSystemOpsBefore;
      this.numNonSystemOpsBefore = 0;
}
```

Feels a bit cleaner to me

vladsud commented on packages/runtime/container-runtime/src/summarizerHeuristics.ts:
```diff
@@ -39,46 +45,72 @@ export class SummarizeHeuristicData implements ISummarizeHeuristicData {
+
+            // We shouldn't attempt a summary if there are no new processed ops
+            const opsSinceLastAck = this.opsSinceLastAck;
+            if (opsSinceLastAck > 0) {
```
Is this check mostly to protect against timer firing when we have no ops?
If so, might be worth to move this check into timer callback, and possible replacing check here with assert - it would more clearly communicate intention and invariants

kian-thompson replied:
Yes, this check is just to prevent ever triggering a summary from this class when we know there are no new ops to summarize.
Moving this check assumes that we only call `ISummarizeHeuristicRunner.run()` after we have processed an op. It's a trade off in my eyes. Do we want to ensure the intention of what happens before this "run" method is called or do we want to always prevent the unnecessary summary?

vladsud commented on packages/runtime/container-runtime/src/summarizerHeuristics.ts:
```diff
@@ -39,46 +45,72 @@ export class SummarizeHeuristicData implements ISummarizeHeuristicData {
+            }
         }
+
+        this.idleTimer?.restart();
```
I think this object is created on demand. That means we already have some trailing ops when we get here, and we already lost ability to separate them into user ops & system ops. We should looking into doing one of the following:
1. Assume some mix (i.e. 100% of user ops, or 70% - 30%) and initiate state appropriately
2. Figure out how to move op counting out of this class, i.e. have some part of this logic be running all the time.
   - That's essentially the same as leveraging OpTracker class. Though we would need to subtract 2 from number of system ops to ignore summarize ops - I think that's fine approach.
    -

vladsud commented on packages/runtime/container-runtime/src/summarizerHeuristics.ts:
```diff
@@ -39,46 +45,72 @@ export class SummarizeHeuristicData implements ISummarizeHeuristicData {
-        this.idleTimer = new Timer(
-            this.configuration.idleTime,
-            () => this.trySummarize("idle"));
+        if (useIdleTimerStrategy) {
```
I personally prefer not to add boolean, but rather use this.configuration.idleTime == Infinity as a way to say timer should not be used. This makes it possible to encode it all in compact way in policy object, and not add more custom code paths / classes to handle such cases.

vladsud commented on packages/runtime/container-runtime/src/summarizerHeuristics.ts:
```diff
@@ -87,6 +119,42 @@ export class SummarizeHeuristicRunner implements ISummarizeHeuristicRunner {
+    }
+}
+
+const DefaultHeuristicWeightConfiguration: ISummarizeHeuristicWeightConfiguration = {
```
I'd prefer not to have these properties be added to ISummaryConfiguration.
Please note that after #9965 (at some future POV in time), we should move ISummaryConfiguration out of protocol-definitions into container runtime. We can do it even now, duplicating interface and starting to use new definition in runtime. I've opened #9990 to partially track that and rationalization of interfaces, but we can do move part now.

vladsud commented on packages/runtime/container-runtime/src/summarizerTypes.ts:
```diff
@@ -95,7 +97,9 @@ export interface ISummarizerRuntime extends IConnectableRuntime {
+    /** @deprecated 1.0, please remove all implementations and usage */
     on(event: "batchEnd", listener: (error: any, op: ISequencedDocumentMessage) => void): this;
+    /** @deprecated 1.0, please remove all implementations and usage */
     removeListener(event: "batchEnd", listener: (error: any, op: ISequencedDocumentMessage) => void): this;
```
I believe this interface is used only internally in this package and should not be exported (if it is exported, your build will fail with back/forward compat tests). I'd rather make sure it's not exported and not care about deprecated markup - we can delete them right away (though see my comment on integrity of batches first).

kian-thompson replied:
If we're exporting the `ContainerRuntime` class don't we also need to export the interfaces it implements? I believe we'd have a visibility mismatch.

kian-thompson committed:
Remove unnecessary heuristic data adjust method

NicholasCouri commented:
>                 (reason) => this.trySummarize(reason));

Trying to catch up here - please forgive me if I'm missing the point :)  

Is there a reason for not starting the runner here ?  

Why do we have the call to 
        summarizer.heuristicRunner?.start();  at all ? 

Why not simply start when/if  creating the SummarizeHeuristicRunner ?   Is there a scenario where we want to start the runner independently ? Why do we have the start at all ? Why not simplify and if we/when we create the idle timer we call   this.idleTimer?.start(); 

---
Refers to: packages/runtime/container-runtime/src/runningSummarizer.ts:125 in f83fb64. [](commit_id = f83fb64d81092c61640854a04ba1ed13e8c2c9e4, deletion_comment = False)

NicholasCouri commented on packages/runtime/container-runtime/src/summarizerHeuristics.ts:
```diff
@@ -87,6 +119,42 @@ export class SummarizeHeuristicRunner implements ISummarizeHeuristicRunner {
     }
 }
+
+export class MaxTimeSummarizeHeuristicStrategy implements ISummarizeHeuristicStrategy {
```
>HeuristicStrategy [](http://example.com/codeflow?start=29&length=17)

nit: Do they need to be called HeuristicStrategy ? Why not simply MaxTimeSummarize and WeightedOpsSummarize

kian-thompson replied:
I named it such since they are dependent on a `ISummarizeHeuristicData` instance. The `Strategy` naming was added because all these classes are used for is to determine if we should run a summary based on some provided state. It should not be doing any processing of the data or running anything important. Doing such allows these strategy classes to be very easily unit testable.

NicholasCouri commented on packages/runtime/container-runtime/src/summarizerHeuristics.ts:
```diff
@@ -87,6 +119,42 @@ export class SummarizeHeuristicRunner implements ISummarizeHeuristicRunner {
+export class MaxTimeSummarizeHeuristicStrategy implements ISummarizeHeuristicStrategy {
+    public readonly summarizeReason: Readonly<SummarizeReason> = "maxTime";
+
+    public shouldRunSummarize(configuration: ISummaryConfiguration, heuristicData: ISummarizeHeuristicData): boolean {
```
>shouldRunSummarize [](http://example.com/codeflow?start=11&length=18)

nit: would it be better shouldRunSummarizer ?

kian-thompson replied:
I agree that the naming of these isn't the nicest. I was following the current naming scheme since the entry method is called `trySummarize` and other things are named similarly (ex: `SummarizeReason`)

NicholasCouri commented:
That helps!  But does it make sense to have or allow the start to happen independently ?  

---
In reply to: [1105496441](https://github.com/microsoft/FluidFramework/pull/9957#issuecomment-1105496441) [](http://example.com/codeflow?ancestors=1105496441,1105496441)

kian-thompson commented:
> That helps! But does it make sense to have or allow the start to happen independently ?
> 
> In reply to: [1105496441](https://github.com/microsoft/FluidFramework/pull/9957#issuecomment-1105496441)

The only way I can see of avoiding any unwanted cases and keeping the `waitStart` functionality would be to set an arbitrary initial timeout. But this is not a guarantee and may also wait an unncessary amount of time. I think as long as we have this `waitStart` method we should start the timer independently :(

kian-thompson commented on packages/runtime/container-runtime/src/runningSummarizer.ts:
```diff
@@ -190,26 +190,14 @@ export class RunningSummarizer implements IDisposable {
-        }
-    }
+    public handleOp(op: ISequencedDocumentMessage) {
+        this.heuristicData.lastOpSequenceNumber = op.sequenceNumber;
```
@NicholasCouri Do you think it's dangerous to not adjust `heuristicData.lastOpSequenceNumber` if the incoming op is something we don't care to summarize? Ex: A `Summarize` op comes in and we skip this method entirely and don't update the sequence number.

NicholasCouri replied:
Not sure - if we skip, the time processing those would not be taken in consideration and I wonder if they are neglectable. - not only that but to debug problems it might be interesting to keep the last sequence up-to-date .   @agarwalnavin / @agarwal-navin  might have more background info :)

agarwal-navin replied:
HeuristicData contains info that is relevant to determining whether we want to do a summary or not. Prior to this change, we did not care about most system ops to make a decision on summaries. Looks like with this change, we do consider system ops as well. So, we will have to track them.

This PR seems complicated and I am not really sure what all its changing. Could you add a description and more comments?
From what I can tell, the gist is that it wants to track how many user vs system ops we have received and make summarize decisions based off of that.

agarwal-navin replied:
`lastOpSequenceNumber` is used to know `opsSinceLastAck`. I am not sure what the meaning of that is anymore - do we mean user + system ops since last ack? or just user ops since last ack?

kian-thompson replied:
`lastOpSequenceNumber` wasn't exactly faithful to its naming in the grand scheme of things. Before, it used to hold the last op sequence number we cared about for summarizing heuristics (i.e. user op batches and the system ops we previously tracked). The system ops we used to care about were `ClientLeave`, `ClientJoin`, and `Propose`. This isn't exactly the best approach as it uses an opt-in mechanism for system ops. This change is trying to better track all the ops that come in before we decide to summarize by providing different weights to different types of ops (user ops = 10x system ops as a base implementation).

NicholasCouri commented on packages/runtime/container-runtime/src/summarizerTypes.ts:
```diff
@@ -394,3 +407,25 @@ type ISummarizeTelemetryOptionalProperties =
+     * @param heuristicData - heuristic data used to confirm conditions are met
+     */
+    shouldRunSummarize(configuration: ISummaryConfiguration, heuristicData: ISummarizeHeuristicData): boolean;
+}
```
New line so the lint compiler won't fail.

NicholasCouri commented on packages/runtime/container-runtime/src/summarizerHeuristics.ts:
```diff
@@ -87,6 +117,42 @@ export class SummarizeHeuristicRunner implements ISummarizeHeuristicRunner {
+        new MaxTimeSummarizeHeuristicStrategy(),
+        new WeightedOpsSummarizeHeuristicStrategy(weightConfiguration),
+    ];
+};
```
>; [](http://example.com/codeflow?start=1&length=1)

new line

NicholasCouri replied:
>; [](http://example.com/codeflow?start=1&length=1)

remove ;

NicholasCouri commented on packages/runtime/container-runtime/src/summarizerHeuristics.ts:
```diff
@@ -5,7 +5,7 @@
 import { Timer } from "@fluidframework/common-utils";
 import { ISummaryConfiguration } from "@fluidframework/protocol-definitions";
-import { ISummarizeHeuristicData, ISummarizeHeuristicRunner, ISummarizeAttempt } from "./summarizerTypes";
+import { ISummarizeHeuristicData, ISummarizeHeuristicRunner, ISummarizeAttempt, ISummarizeHeuristicStrategy, ISummarizeHeuristicWeightConfiguration } from "./summarizerTypes";
```
break the line (<120)  - it is complaining on your PR :)

NicholasCouri commented on packages/runtime/container-runtime/src/summarizerHeuristics.ts:
```diff
@@ -39,46 +45,70 @@ export class SummarizeHeuristicData implements ISummarizeHeuristicData {
+            if (opsSinceLastAck > 0) {
+                trySummarize(reason);
+            }
+        }
```
>} [](http://example.com/codeflow?start=8&length=1)

}; (from PR)

NicholasCouri commented on packages/runtime/container-runtime/src/summarizerHeuristics.ts:
```diff
@@ -87,6 +117,42 @@ export class SummarizeHeuristicRunner implements ISummarizeHeuristicRunner {
+const DefaultHeuristicWeightConfiguration: ISummarizeHeuristicWeightConfiguration = {
+    systemOpWeight   : 0.1,
+    nonSystemOpWeight: 1.0,
+}
```
};

agarwal-navin commented on packages/runtime/container-runtime/src/summarizerHeuristics.ts:
```diff
@@ -20,6 +20,12 @@ export class SummarizeHeuristicData implements ISummarizeHeuristicData {
     }
 
+    public numSystemOps: number = 0;
+    private numSystemOpsBefore: number = 0;
```
What do the before fields mean? Could you please add a comment.

kian-thompson replied:
These variables store the op counts at the beginning of a summarization attempt. They are used to adjust the op counts appropriately once a summarization is completed.

kian-thompson committed:
Fix some linting issues

kian-thompson committed:
Add additional comment to member variables

vladsud commented on packages/runtime/container-runtime/src/summarizerHeuristics.ts:
```diff
@@ -39,46 +59,70 @@ export class SummarizeHeuristicData implements ISummarizeHeuristicData {
+            }
         }
+
+        this.idleTimer?.restart();
```
I mentioned it somewhere before, but I do not think it's addressed. The code before would not call run() method for summarize and summarize ack ops. This would mean we would not be summarizing these ops in a loop infinitely.

Now, given that we call run() for all system ops, this area is regressed. I'd love to see a UT that catches that case, and a fix. Fix can be either not call run for those two ops from handleOp() handlers. Maybe there are other ways to avoid it.

vladsud commented on packages/runtime/container-runtime/src/summarizerHeuristics.ts:
```diff
@@ -39,46 +59,70 @@ export class SummarizeHeuristicData implements ISummarizeHeuristicData {
     }
 
+    public start() {
+        this.idleTimer?.start();
```
I think it should instead call run() method. Two reasons:
- If we have some ops accumulated already, we should consider to summarize them
- If we have no ops accumulated, we should not set timer (as we should not summarize nothing).

That said, it's correct if we loaded from latest summary.
In such case our op counting would be correct.
I'm not sure how code works today if it's not the case. I think we have no way to accurately collect data about user & system ops, we can only see total ops (as we learn about reference sequence number of a summary too late).
I do not even know if it worked correctly before your changes.
Let's chat about it.

kian-thompson commented on packages/runtime/container-runtime/src/summarizer.ts:
```diff
@@ -277,11 +276,8 @@ export class Summarizer extends EventEmitter implements ISummarizer {
-
-        this.opListener = (error: any, op: ISequencedDocumentMessage) => runningSummarizer.handleOp(error, op);
-        this.runtime.on("batchEnd", this.opListener);
+        this.opListener = (op: ISequencedDocumentMessage) => runningSummarizer.handleOp(op);
```
@vladsud I thought about how we may move this handler logic to before we construct the object, but I don't think I like the idea. It would have to be very reliant on how this `Summarizer.start(...)` method and further up is called (because of asynchronous nature), so it may make a lot more sense to have this backtracking logic handled by the `RunningSummarizer` after we call `waitStart`.

For example:
- We initialize the `RunningSummarizer` object and wait for it to start
- Provide the `RunningSummarizer` some more up-to-date op sequence number and pass that along to the heuristic data appropriately
- Tell the heuristic runner to start and also run an initial heuristic check

Is there somewhere we can get an up-to-date op sequence number easily? Or will we have to do implement something new to track this and then pass the information along to the `RunningSummarizer`?

vladsud replied:
It's trivial to get number of ops (as we know latest op we processed, and reference sequence number of a summary we loaded from). That's what existing code (before your changes) does when deciding when to run a summary.
We will lose break down of info on user vs. system ops though in that way. It may be fine, as we can guess / do some allocation, i.e. assume that it's 50% - 50% or some other rate that makes sense.

kian-thompson replied:
Yes, we'd probably have to assume some basic op ratio or potentially do a different check altogether. But currently we only update the current op sequence number in the op handler for the `RunningSummarizer`. My understanding is that there may be some ops we'd like to process a summary for in the following cases:
1. Op is processed during the time that we are initiating the `RunningSummarizer` instance but before we listen for the op events (will get missed by the handlers in the current workflow)
2. Op was sequenced after the last time we summarized (op sequence number > summarize ref sequence number)

So, is there currently a good way to retrieve the most recent op sequence number (so we can run heuristics) before we update the value for `this.heuristicData.lastOpSequenceNumber`?

vladsud replied:
lastOpSequenceNumber looks very weird.
Yes, you can always get it from IDeltaManager.lastSequenceNumber, which is same as IDeltaManager.lastMessage.sequenceNumber.

Summarizer has access to delta manager through this.runtime.deltaManager, I have not looked if it's propagated through the layers.

You actually see it used here, where we populate initial value:
```ts
        const runningSummarizer = await RunningSummarizer.start(
            this.logger,
            this.summaryCollection.createWatcher(clientId),
            this.configurationGetter(),
            async (...args) => this.internalsProvider.submitSummary(...args), // submitSummaryCallback
            new SummarizeHeuristicData(
                this.runtime.deltaManager.lastSequenceNumber,
                { /** summary attempt baseline for heuristics */
                    refSequenceNumber: this.runtime.deltaManager.initialSequenceNumber,
                    summaryTime: Date.now(),
                } as const,
            ),
```

Are you sure there is no way to move initialization of handler up in time? :)
I think you will need to pass runtime down to RunningSummarizer either way, and if so, we can move responsibility to register for events into RunningSummarizer. I think it's more natural flow - it feels weird that Summarizer.start() does this event registration for RunningSummarizer

kian-thompson commented on packages/runtime/container-runtime/src/summarizer.ts:
```diff
@@ -277,11 +276,8 @@ export class Summarizer extends EventEmitter implements ISummarizer {
-
-        this.opListener = (error: any, op: ISequencedDocumentMessage) => runningSummarizer.handleOp(error, op);
-        this.runtime.on("batchEnd", this.opListener);
+        this.opListener = (op: ISequencedDocumentMessage) => runningSummarizer.handleOp(op);
```
@vladsud I thought about how we may move this handler logic to before we construct the object, but I don't think I like the idea. It would have to be very reliant on how this `Summarizer.start(...)` method and further up is called (because of asynchronous nature), so it may make a lot more sense to have this backtracking logic handled by the `RunningSummarizer` after we call `waitStart`.

For example:
- We initialize the `RunningSummarizer` object and wait for it to start
- Provide the `RunningSummarizer` some more up-to-date op sequence number and pass that along to the heuristic data appropriately
- Tell the heuristic runner to start and also run an initial heuristic check

Is there somewhere we can get an up-to-date op sequence number easily? Or will we have to do implement something new to track this and then pass the information along to the `RunningSummarizer`?

kian-thompson committed:
Implement handling of missed ops
- TODO fix relevant unit tests

kian-thompson committed:
Prevent running summaries on certain op types

vladsud commented on packages/runtime/container-runtime/src/runningSummarizer.ts:
```diff
@@ -68,12 +71,34 @@ export class RunningSummarizer implements IDisposable {
+            heuristicData.lastSuccessfulSummary.refSequenceNumber
+            + heuristicData.numSystemOps
+            + heuristicData.numNonSystemOps);
+        if (diff > 0) {
```
can diff be negative here?

kian-thompson replied:
It's not possible based on my understanding of what these properties are tracking. In any case, if the number is negative, we probably can't meaningfully increase the counters.

vladsud commented on packages/runtime/container-runtime/src/runningSummarizer.ts:
```diff
@@ -68,12 +71,34 @@ export class RunningSummarizer implements IDisposable {
+        if (diff > 0) {
+            // Split the diff 50-50 and increment the counts appropriately
+            heuristicData.numSystemOps += Math.ceil(diff / 2);
+            heuristicData.numNonSystemOps += Math.floor(diff / 2);
```
minor - off-by-1 error due to rounding of odd/2 calculus

kian-thompson replied:
Wouldn't `Math.ceil` and `Math.floor` catch this case? From what I remember, Javascript automatically includes the decimals since there is only the `number` type.

vladsud replied:
looks good in terms of math.

vladsud commented on packages/runtime/container-runtime/src/runningSummarizer.ts:
```diff
@@ -190,30 +221,45 @@ export class RunningSummarizer implements IDisposable {
+     * @param op - op to check
+     * @returns true if this type of op can trigger a summary
+     */
+    private opCanTriggerSummary(op: ISequencedDocumentMessage): boolean {
```
I'd rather exclude them from heuristicData.numSystemOps - I think it's better place to do so, seems cleaner.

kian-thompson replied:
I'm not sure if I'm a fan of this approach. The only thing that would change is potentially just hitting the `maxTime` summary reason instead of `idleTime`. But it has the cost of the heuristic data no longer being true to its naming and being a bit harder to follow.

vladsud commented on packages/runtime/container-runtime/src/runningSummarizer.ts:
```diff
@@ -264,6 +310,7 @@ export class RunningSummarizer implements IDisposable {
                 summarySequenceNumber: waitStartResult.value.summaryOp.sequenceNumber,
             });
         }
+        this.readyToStartSummarizing = true;
```
I'd rename it to something like initialized.
BTW, why we did not need this kind of check before?

kian-thompson replied:
It was not needed before because we set up the handlers after awaiting `RunningSummarizer.start`. Now, we listen to the op events before we could potentially want to run a heuristic summary.

kian-thompson committed:
Rename variable to initialized

kian-thompson committed:
Merge branch 'next' into 9922-weight-system-ops-differently

kian-thompson committed:
Remove unnecessary parameter

kian-thompson committed:
Merge branch 'next' into 9922-weight-system-ops-differently

kian-thompson committed:
Merge from next

kian-thompson committed:
Move weight configuration up to new ISummaryConfigurationHeuristics interface

kian-thompson committed:
Remove ISummarizeHeuristicWeightConfiguration

kian-thompson committed:
Adjust naming of heuristic strategy classes and interfaces

kian-thompson committed:
Add some comments

kian-thompson committed:
Pass mock runtime in tests

kian-thompson committed:
Fix summarizer heuristic tests

kian-thompson committed:
Fix linting errors

kian-thompson committed:
Standardize format for default summary configuration

kian-thompson committed:
Reduce maxOps to 100

kian-thompson committed:
Adjust comment

kian-thompson committed:
API changes to container runtime

kian-thompson committed:
Fix tests

kian-thompson committed:
Add unit test for summary ops

kian-thompson committed:
Fix e2e tests
- Give both system and non-system ops the same weight so there's no guessing on what type of ops will be used

kian-thompson committed:
Unit test weighting functionality

NicholasCouri commented on packages/runtime/container-runtime/src/containerRuntime.ts:
```diff
@@ -252,21 +260,25 @@ export type ISummaryConfiguration =
     minOpsForLastSummaryAttempt: 50,
 
-    maxAckWaitTime: 6 * 10 * 1000, // 6 min.
+    maxAckWaitTime: 6 * 10 * 1000, // 60 mins.
```
I think we messed this up - it used to 600,000 
and it is now only 60 seconds (not 60 mins) 

Unless  Vlad asked to change it, it should be 600 * 1000  600 seconds or  10 mins

kian-thompson committed:
Revert change to maxAckWaitTime

NicholasCouri reviewed (approved):
:shipit:

kian-thompson committed:
Merge branch 'next' into 9922-weight-system-ops-differently

vladsud commented on packages/runtime/container-runtime/src/containerRuntime.ts:
```diff
@@ -228,12 +228,20 @@ export interface ISummaryConfigurationHeuristics extends ISummaryBaseConfigurati
      */
     minOpsForLastSummaryAttempt: number;
+    /**
+     * System op weight to use in heuristic summarizing
```
might be useful to provide some comment about expected range and how it will be used. Otherwise, it's not very clearly how to use it. For example, is using 1.0 & 1.0 for weights same as 2.0 & 2.0 or not? Do they need to add up to 1.0 together? Things like that are useful to mention (and enforce)

vladsud commented on packages/runtime/container-runtime/src/containerRuntime.ts:
```diff
@@ -252,21 +260,25 @@ export type ISummaryConfiguration =
+
+    systemOpWeight: 0.1,
+
+    nonSystemOpWeight: 1.0,
```
Should we call it user op weight?
I think we plan to use "runtime" word - see https://github.com/microsoft/FluidFramework/issues/9732

vladsud replied:
Let's find a way to use same terminology everywhere.

kian-thompson replied:
Are you suggesting to use `clientOpWeight` and `runtimeOpWeight` instead of `systemOpWeight` and `nonSystemOpWeight` respectively?

vladsud replied:
I think it's fine to use runtime and non-runtime pair, or runtime & system pair of words. While later does not exactly map to actual terms, I think it's fine.

vladsud commented on packages/runtime/container-runtime/src/runningSummarizer.ts:
```diff
@@ -196,30 +227,46 @@ export class RunningSummarizer implements IDisposable {
+    public handleOp(op: ISequencedDocumentMessage) {
+        this.heuristicData.lastOpSequenceNumber = op.sequenceNumber;
+
+        if (isSystemMessage(op)) {
```
Let's use here isRuntimeMessage, or even better - just op.type === "op" (i.e. not count summarize op as runtime op).
Not sure if we want to have a funciton for that.

vladsud reviewed (approved):
looks good - please take a look at old & new comments.
Please make sure to test it manually (across all interesting combinations) and also see if we have any gaps in UT coverage, especially around things like not summarizing summary ops, triggering summary on all triggers (idle, maxtime, max ops).

kian-thompson committed:
Merge branch 'next' into 9922-weight-system-ops-differently

kian-thompson committed:
Change naming of properties and add unit test

kian-thompson committed:
Add better comment to op weights

kian-thompson committed:
Reorganize unit test

kian-thompson committed:
Remove unused import

kian-thompson committed:
Fix lint

kian-thompson committed:
Fix API change
